File: python/examples/all_classifier.py
#!/usr/bin/env python
"""
Explicit examples on how to use the different classifiers
"""

from numpy import double, array, floor, concatenate, sign, ones, zeros
from numpy.random import rand, seed, permutation
from sg import sg

def get_clouds (num, num_feats, num_vec):
	data=[rand(num_feats, num_vec)+x/2 for x in xrange(num)]
	cloud=concatenate(data, axis=1)
	return array([permutation(x) for x in cloud])

def get_dna ():
	acgt=array(['A', 'C', 'G','T'])
	len_acgt=len(acgt)
	rand_train=[]
	rand_test=[]

	for i in xrange(11):
		str1=[]
		str2=[]
		for j in range(60):
			str1.append(acgt[floor(len_acgt*rand())])
			str2.append(acgt[floor(len_acgt*rand())])
		rand_train.append(''.join(str1))
	rand_test.append(''.join(str2))
	
	for i in xrange(6):
		str1=[]
		for j in range(60):
			str1.append(acgt[floor(len_acgt*rand())])
	rand_test.append(''.join(str1))

	return {'train': rand_train, 'test': rand_test}

###########################################################################
# kernel-based SVMs
###########################################################################

def svm_light ():
	print 'SVMLight'

	data=get_dna()
	size_cache=10
	degree=20
	C=0.017
	epsilon=1e-5
	use_bias=False
	num_feats=14
	num_trainvec=11
	labels=sign(rand(1, num_trainvec)-0.5)[0]

	sg('set_features', 'TRAIN', data['train'], 'DNA')
	sg('set_kernel', 'WEIGHTEDDEGREE', 'CHAR', size_cache, degree)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', labels)
	sg('new_svm', 'LIGHT')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', data['test'], 'DNA')
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

def libsvm ():
	print 'LibSVM'

	size_cache=10
	width=2.1
	C=0.017
	epsilon=1e-5
	use_bias=False
	num_feats=11
	num_trainvec=12

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 17)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', trainlab)
	sg('new_svm', 'LIBSVM')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

def gpbtsvm ():
	print 'GPBTSVM'

	size_cache=10
	width=2.1
	C=0.017
	epsilon=1e-5
	use_bias=False
	num_feats=9
	num_trainvec=11

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 16)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', trainlab)
	sg('new_svm', 'GPBTSVM')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

def mpdsvm ():
	print 'MPDSVM'

	size_cache=10
	width=2.1
	C=0.017
	epsilon=1e-5
	use_bias=False
	num_feats=13
	num_trainvec=11

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 18)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', trainlab)
	sg('new_svm', 'MPDSVM')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

def libsvm_multiclass ():
	print 'LibSVMMultiClass'

	size_cache=10
	width=2.1
	C=10.
	epsilon=1e-5
	use_bias=False
	num_feats=9
	num_trainvec=11

	trainlab=array([double(x) for x in xrange(num_trainvec*2)])
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 23)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', trainlab)
	sg('new_svm', 'LIBSVM_MULTICLASS')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

def libsvm_oneclass ():
	print 'LibSVMOneClass'

	size_cache=10
	width=2.1
	C=10.
	epsilon=1e-5
	use_bias=False
	num_feats=9
	num_trainvec=11

	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 23)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('new_svm', 'LIBSVM_ONECLASS')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

def gmnpsvm ():
	print 'GMNPSVM'

	size_cache=10
	width=2.1
	C=0.017
	epsilon=1e-5
	use_bias=False
	num_feats=32
	num_trainvec=12

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 18)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', trainlab)
	sg('new_svm', 'GMNPSVM')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')
	result=sg('svm_classify')

###########################################################################
# run with batch or linadd on LibSVM
###########################################################################

def do_batch_linadd ():
	print 'LibSVM batch'

	size_cache=10
	width=2.1
	C=0.017
	epsilon=1e-5
	use_bias=False
	num_feats=11
	num_trainvec=12

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 17)

	sg('set_features', 'TRAIN', traindata)
	sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
	sg('init_kernel', 'TRAIN')

	sg('set_labels', 'TRAIN', trainlab)
	sg('new_svm', 'LIBSVM')
	sg('svm_epsilon', epsilon)
	sg('c', C)
	sg('svm_use_bias', use_bias)
	sg('svm_train')

	sg('set_features', 'TEST', testdata)
	sg('init_kernel', 'TEST')

	objective=sg('get_svm_objective')
	sg('use_batch_computation', True)
	sg('use_linadd', True)
	result=sg('svm_classify')

###########################################################################
# misc classifiers
###########################################################################

def perceptron ():
	print 'Perceptron'

	num_feats=14
	num_trainvec=10

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 42)

	sg('set_features', 'TRAIN', traindata)
	sg('set_labels', 'TRAIN', trainlab)
	sg('new_classifier', 'PERCEPTRON')
	sg('train_classifier')

	sg('set_features', 'TEST', testdata)
	result=sg('classify')

def knn ():
	print 'KNN'

	num_feats=14
	num_trainvec=10
	k=3

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 42)

	sg('set_features', 'TRAIN', traindata)
	sg('set_labels', 'TRAIN', trainlab)
	sg('set_distance', 'EUCLIDIAN', 'REAL')
	sg('init_distance', 'TRAIN')
	sg('new_classifier', 'KNN')
	sg('train_classifier', k)

	sg('set_features', 'TEST', testdata)
	sg('init_distance', 'TEST')
	result=sg('classify')

def lda ():
	print 'LDA'

	num_feats=14
	num_trainvec=10

	trainlab=sign(rand(1, num_trainvec*2)-0.5)[0]
	traindata=get_clouds(2, num_feats, num_trainvec)
	testdata=get_clouds(2, num_feats, 42)

	sg('set_features', 'TRAIN', traindata)
	sg('set_labels', 'TRAIN', trainlab)
	sg('new_classifier', 'LDA')
	sg('train_classifier')

	sg('set_features', 'TEST', testdata)
	result=sg('classify')

###########################################################################
# call functions
###########################################################################

if __name__=='__main__':
	seed(42)

	svm_light()
	libsvm()
	gpbtsvm()
	mpdsvm()
	libsvm_multiclass()
	libsvm_oneclass()
	gmnpsvm()

	do_batch_linadd()

	perceptron()
	knn()
	lda()

File: r/examples/all_classifier.R
# Explicit examples on how to use the different classifiers
#
# run as R --no-save --slave --file=<filename>

library("sg")
#uncomment if make install does not work and comment the library("sg") line above
#dyn.load('sg.so')
#sg <- function(...) .External("sg",...,PACKAGE="sg")

len <- 12
num <- 30
size_cache <- 10
C <- 10
epsilon <- 1e-5
use_bias <- TRUE

traindat_real <- matrix(c(rnorm(len*num)-1,rnorm(len*num)+1), len, 2*num)
testdat_real <- matrix(c(rnorm(len*num)-1,rnorm(len*num)+1), len, 2*num)
trainlab_one <- c(rep(-1,num),rep(1,num))
trainlab_multi <- c(rep(0,num/2), rep(1,num/2), rep(2,num/2), rep(3,num/2))


#
# kernel-based SVMs
#

getDNA <- function(len, num) {
	acgt <- c('A', 'C', 'G', 'T')
	data <- c()
	for (j in 1:num) {
		str <- '';
		for (i in 1:len) {
			str <- paste(str, sample(acgt, 1), sep='')
		}
		data <- append(data, str)
	}
	data
}

trainlab_dna <- c(rep(-1,num/2),rep(1, num/2))
traindat_dna <- getDNA(len, num)
testdat_dna <- getDNA(len, num+7)

degree <- 20

# SVM Light
dosvmlight <- function()
{
	print('SVMLight')

	dump <- sg('set_features', 'TRAIN', traindat_dna, 'DNA')
	dump <- sg('set_kernel',  'WEIGHTEDDEGREE', 'CHAR', size_cache, degree)
	dump <- sg('init_kernel', 'TRAIN')

	dump <- sg('set_labels', 'TRAIN', trainlab_dna)

	dump <- sg('new_svm', 'LIGHT')
	dump <- sg('svm_epsilon', epsilon)
	dump <- sg('c', C)
	dump <- sg('svm_use_bias', use_bias)
	dump <- sg('train_classifier')

	dump <- sg('set_features', 'TEST', testdat_dna, 'DNA')
	dump <- sg('init_kernel', 'TEST')
	result <- sg('classify')
}
try(dosvmlight())


# LibSVM
print('LibSVM')

width <- 2.1

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_svm', 'LIBSVM')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')
result <- sg('classify')


# GPBTSVM
print('GPBTSVM')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_svm', 'GPBTSVM')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')
result <- sg('classify')


# MPDSVM
print('MPDSVM')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_svm', 'MPDSVM')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')
result <- sg('classify')


# LibSVM MultiClass
print('LibSVMMultiClass')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_multi)
dump <- sg('new_svm', 'LIBSVM_MULTICLASS')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')
result <- sg('classify')


# LibSVMOneClass
print('LibSVMOneClass')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_svm', 'LIBSVM_ONECLASS')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')
result <- sg('classify')


# GMNPSVM
print('GMNPSVM')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_svm', 'GMNPSVM')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')
result <- sg('classify')


#
# run with batch or linadd on LibSVM
#

# LibSVM batch
print('LibSVM batch')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_kernel', 'GAUSSIAN', 'REAL', size_cache, width)
dump <- sg('init_kernel', 'TRAIN')

dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_svm', 'LIBSVM')
dump <- sg('svm_epsilon', epsilon)
dump <- sg('c', C)
dump <- sg('svm_use_bias', use_bias)
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_kernel', 'TEST')

objective <- sg('get_svm_objective')
dump <- sg('use_batch_computation', TRUE)
dump <- sg('use_linadd', TRUE)
result <- sg('classify')


#
# misc classifiers
#

# Perceptron
print('Perceptron')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_classifier', 'PERCEPTRON')
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
result <- sg('classify')


# KNN
print('KNN')
k <- 3

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('set_distance', 'EUCLIDIAN', 'REAL')
dump <- sg('init_distance', 'TRAIN')
dump <- sg('new_classifier', 'KNN')
dump <- sg('train_classifier', k)

dump <- sg('set_features', 'TEST', testdat_real)
dump <- sg('init_distance', 'TEST')
result <- sg('classify')


# LDA
print('LDA')

dump <- sg('set_features', 'TRAIN', traindat_real)
dump <- sg('set_labels', 'TRAIN', trainlab_one)
dump <- sg('new_classifier', 'LDA')
dump <- sg('train_classifier')

dump <- sg('set_features', 'TEST', testdat_real)
result <- sg('classify')

